File: Device_control/PI-E709/PI_codes/PI-Software-Suite-C-990.CD1/Development/C#/Samples/General Controller/AutoZeroSample/AutoZeroSample.cs
/**  <copyright file="AutoZero.cs" company="Physik Instrumente (PI) SE & Co. KG">
  *  (c)2015-2020 Physik Instrumente (PI) SE & Co. KG
  *  Software products that are provided by PI are subject to the General Software License Agreement of
  *  Physik Instrumente (PI) SE & Co. KG and may incorporate and/or make use of third-party software components.
  *  For more information, please read the General Software License Agreement and the Third Party Software Note linked below.
  *
  *  General Software License Agreement :
  *  http://www.physikinstrumente.com/download/EULA_PhysikInstrumenteGmbH_Co_KG.pdf
  *  Third Party Software Note :
  *  http://www.physikinstrumente.com/download/TPSWNote_PhysikInstrumenteGmbH_Co_KG.pdf
  *  </copyright>* /

/** <remarks>
 *  !! MOTION WARNING !!
 *
 *  This sample connects to a controller (i.e. E-727) via connection dialog
 *  and which supports auto zero.
 *  The controller will then do a auto zero of the stage.
 *  The controller and the stage has to be initialized with PIMikroMove before starting this sample.
 *  </remarks>*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using PI;

namespace AutoZeroSample
{
    public enum ConnectionType
    {
        Dialog,
        Rs232,
        Tcpip,
        Usb
    }

    public struct AxisState
    {
        public string Axis;
        public int State;

        public AxisState(string axis, int state)
        {
            Axis = axis;
            State = state;
        }
    }

    /**  <summary>
      *  This executes a auto zero for controller which support this feature i.e E-727.
      *  Call the Run-method to execute this sample.
      *  </summary>*/
    public class AutoZeroSample
    {
        #region controller specific configuration
        public struct ReferenceSettings
        {
            internal string AxisIdentifier { get; }

            internal int[] ServoState { get; }

            public ReferenceSettings(string axisAxisIdentifier, int[] servoState)
            {
                AxisIdentifier = axisAxisIdentifier;
                ServoState = servoState;
            }
        }

        private static readonly ReferenceSettings Settings
            = new ReferenceSettings("1", new []{PI_FALSE});
        #endregion controller specific configuration

        #region fields
        private const int PI_RESULT_FAILURE = 0;
        private const int PI_TRUE  = 1;
        private const int PI_FALSE = 0;
        internal int ControllerId = -1;
        #endregion fields

        #region public methods
        public void Run()
        {
            try
            {
                OpenConnection(ConnectionType.Dialog);
                PrintControllerIdentification();

                SetServoState(Settings.AxisIdentifier, Settings.ServoState);
                AutoZeroStage(Settings.AxisIdentifier);
                SetServoState(Settings.AxisIdentifier, new [] {PI_FALSE});

            }
            catch (GcsCommandError e)
            {
                PrintErrorMessage(e);
            }
            finally
            {
                DisconnectController();
            }
        }
        #endregion

        #region private methods
        private void AutoZeroStage(string axisIdentifier)
        {
            var axisIdentifierSplit = axisIdentifier.Split(' ');
            var autoZeroState = Enumerable.Repeat(PI_TRUE, axisIdentifierSplit.Length).ToArray();

            // check auto zero state
            if ( GCS2.qATZ(ControllerId, axisIdentifier, autoZeroState) == PI_RESULT_FAILURE)
            {
                throw new GcsCommandError("Unable to check auto zero state.");
            }

            var axisEnumerableStates = GetAxisEnumerableStates(axisIdentifierSplit, autoZeroState, PI_TRUE);

            // start auto zero procedure
            if ( axisEnumerableStates.Any(axisState
                => GCS2.ATZ(ControllerId, axisState.Axis, null, new []{axisState.State})
                   == PI_RESULT_FAILURE))
            {
                throw new GcsCommandError("Unable to start auto zero.");
            }

            // Wait until the auto zero is done.
            var bRefDone = PI_FALSE;

            Console.WriteLine("Waiting for auto zero to be finished.");
            while ( bRefDone == PI_FALSE )
            {
                if (GCS2.IsControllerReady(ControllerId,ref bRefDone ) == PI_RESULT_FAILURE)
                {
                    throw new GcsCommandError("Unable to retrieve controller ready state.");
                }
            }

            if ( GCS2.qATZ(ControllerId, axisIdentifier, autoZeroState) == PI_RESULT_FAILURE)
            {
                throw new GcsCommandError("Unable to check auto zero state.");
            }

            axisEnumerableStates = GetAxisEnumerableStates(axisIdentifierSplit, autoZeroState, PI_TRUE);

            foreach (var axisStateTrue in axisEnumerableStates)
            {
                Console.WriteLine("Settings " + axisStateTrue.Axis + " has done auto zero.");
            }

            axisEnumerableStates = GetAxisEnumerableStates(axisIdentifierSplit, autoZeroState, PI_TRUE);

            foreach (var axisStateFalse in axisEnumerableStates)
            {
                Console.WriteLine("Settings " + axisStateFalse.Axis + " has not done auto zero.");
            }

        }

        private IEnumerable<AxisState> GetAxisEnumerableStates(string[] axisIdentifierSplit, IEnumerable<int> autoZeroState, int filterState)
        {
            return from state in axisIdentifierSplit
                    .Zip(autoZeroState, (axis, state) => new AxisState(axis, state))
                where state.State == filterState
                select state;
        }

        private void SetServoState(string axis, int[] state)
        {
            if (GCS2.SVO(ControllerId, axis, state ) == PI_RESULT_FAILURE )
            {
                throw new GcsCommandError("SVO failed. Exiting");
            }
        }

        private void OpenConnection(ConnectionType type, string hostName = "localhost")
        {
            Console.WriteLine("-- Try to Connect --");
            switch (type)
            {
                case ConnectionType.Dialog:
                    ControllerId = GCS2.InterfaceSetupDlg("");
                    break;
                case ConnectionType.Rs232:
                    ControllerId = GCS2.ConnectRS232(1, 115200);
                    break;
                case ConnectionType.Tcpip:
                    ControllerId = GCS2.ConnectTCPIP(hostName, 50000);
                    break;
                case ConnectionType.Usb:
                {
                    var connectedUsbController = new StringBuilder(1024);
                    var noDevices = GCS2.EnumerateUSB(connectedUsbController, 1024, "");

                    Console.WriteLine("Found: " + noDevices + " USB controllers, connecting to the first one:");
                    Console.Write(connectedUsbController + "\n");

                    ControllerId = GCS2.ConnectUSB(
                        connectedUsbController
                            .ToString()
                            .Split(new[] {Environment.NewLine}, StringSplitOptions.None)
                            [0]);
                    break;
                }
                default:
                    throw new ArgumentOutOfRangeException(nameof(type), type, null);
            }

            if (0 <= ControllerId) return;

            var connType = type.Equals(ConnectionType.Dialog) ? type.ToString() : type.ToString().ToUpper();
            throw new GcsCommandError("Connection to " + connType + " failed. Exiting.");
        }

        private void DisconnectController()
        {
            if (ControllerId >= 0)
            {
                GCS2.CloseConnection(ControllerId);
                ControllerId = -1;
            }

            Console.WriteLine("Close connection.");
        }

        private void PrintErrorMessage(GcsCommandError e)
        {
            var errorCode = GCS2.GetError(ControllerId);
            var errorMessage = new StringBuilder(1024);
            GCS2.TranslateError(errorCode, errorMessage, errorMessage.Capacity);
            Console.WriteLine("ERROR: " + e.Message + " - Reported error code: " + errorCode + ": " + errorMessage);
        }

        private void PrintControllerIdentification()
        {
            var controllerIdentification = new StringBuilder(1024);

            if (GCS2.qIDN(ControllerId, controllerIdentification, controllerIdentification.Capacity) ==
                PI_RESULT_FAILURE)
            {
                throw (new GcsCommandError("qIDN failed. Exiting."));
            }

            Console.WriteLine("qIDN returned: " + controllerIdentification);
        }
        #endregion
    }

    internal class GcsCommandError : Exception
    {
        public GcsCommandError( string message )
            : base(message)
        {
        }
    }
}

File: Device_control/PI-E709/PI_codes/PI-Software-Suite-C-990.CD1/Development/MATLAB/Samples/General Controller/AutoZeroSample/AutoZeroSample.m
%   (c)2015-2020 Physik Instrumente (PI) SE & Co. KG
%   Software products that are provided by PI are subject to the General Software License Agreement of 
%   Physik Instrumente (PI) SE & Co. KG and may incorporate and/or make use of third-party software components. 
%   For more information, please read the General Software License Agreement and the Third Party Software Note linked below.
% 
%   General Software License Agreement :
%   http://www.physikinstrumente.com/download/EULA_PhysikInstrumenteGmbH_Co_KG.pdf
%   Third Party Software Note :
%   http://www.physikinstrumente.com/download/TPSWNote_PhysikInstrumenteGmbH_Co_KG.pdf
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%   IMPORTANT NOTICE
%
%   This sample only works with a PI_MATLAB_Driver version 1.5.0 and newer.
% 
%   You must always initialize the system first. The steps necessary depend on the type of mechanics (e.g., positioner, actuator) 
%   and electronics (e.g., controller, drive electronics). 
%   It may be necessary to load positioner parameters from a database and/or perform a homing procedure (e.g., auto zero procedure, reference move).
%   Refer to the manuals of your mechanics and electronics for details.
%
%   For the initialization use PIMikroMove on Windows or PITerminal on Linux.
%
%   - Load positioner parameters via "Start Up Controller" in PIMikroMove 
%   or via the CST command in PITerminal. 
%
%   - Execute the homing procedure in PIMikroMove via "Start Up Controller" -> "Start up axes".
%   Alternatively, use the corresponding sample (AutoZeroSample, ReferenceSample, ReferenceWithEAXSample). 
%
%   To check if the PI_MATLAB_Driver works properly, run the sample 
%   BeginWithThisSampleForTestingMatlabDriver.m
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%  !! MOTION WARNING !!
%
%   Description:
%
%   This sample connects to a controller which supports auto zero (e.g. E-727) with a positioner connected.
%   Then the controller runs an auto zero procedure to set the zero position of the axis.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Setup configuration

% You MUST EDIT AND ACTIVATE the parameters to make your system run
% properly. 
% 1. Activate the connection type
% 2. Set the connection settings

% Connection type
use_RS232_Connection                = false;
use_TCPIP_Connection                = true;
use_USB_Connection                  = false;

% Connection settings
% RS232
comPort = 1;          % Look at the Device Manager to get the correct COM Port
baudRate = 115200;    % Look at the manual to get the correct baud rate for your controller
% USB
controllerSerialNumber = 'Enter valid serial number here, e.g. "123456789"'; % Use "devicesUsb = Controller.EnumerateUSB('')" to get all PI controller connected to your PC. Or look at the label of the case of your controller
% TCP/IP
ip = 'XXX.XXX.XXX.XX';  % Use "devicesTcpIp = Controller.EnumerateTCPIPDevices('')" to get all PI controller available on the network
port = 50000;           % Is 50000 for almost all PI controllers

%% Load PI MATLAB Driver GCS2
isWindows   = any (strcmp (mexext, {'mexw32', 'mexw64'}));

if(~isdeployed) % Determine whether code is running in deployed or MATLAB mode
    if (isWindows)
        addpath (getenv ('PI_MATLAB_DRIVER'));
    else
        addpath ( '/usr/local/PI/pi_matlab_driver_gcs2' );
    end
end

% Load PI_GCS_Controller if not already loaded
if(~exist('Controller','var'))
    Controller = PI_GCS_Controller();
end
if(~isa(Controller,'PI_GCS_Controller'))
    Controller = PI_GCS_Controller();
end

%% Start connection
% (if not already connected)
try
    boolPIdeviceConnected = false; 
    if ( exist ( 'PIdevice', 'var' ) ), if ( PIdevice.IsConnected ), boolPIdeviceConnected = true; end; end
    if ( ~(boolPIdeviceConnected ) )

    % RS232
        if (use_RS232_Connection)
            if (isWindows)
                PIdevice = Controller.ConnectRS232 ( comPort, baudRate );
            else
                DeviceName = '/dev/ttyS0';  % Select the serial device
                PIdevice = Controller.ConnectRS232ByDevName ( DeviceName, baudRate );
            end
        end

    % USB
        if(use_USB_Connection)
            PIdevice = Controller.ConnectUSB ( controllerSerialNumber );
        end
    % TCP/IP
        if(use_TCPIP_Connection)
            PIdevice = Controller.ConnectTCPIP ( ip, port );
        end

    end

    % query controller identification string
    connectedControllerName = PIdevice.qIDN();

    % initialize PIdevice object for use in MATLAB
    PIdevice = PIdevice.InitializeController ();


    %% Auto Zero axis
    axis = '1';
    

    PIdevice.SVO(axis, 1);

    % set axes to zero position
    PIdevice.ATZ(axis)  

    % wait until all axes are set to zero: all([1 1 1]) >> 1    
    while(0 == PIdevice.qATZ(axis))                               
        pause(0.1);
        fprintf('.')
    end

    disp('AutoZero Complete');
    %% If you want to close the connection
    PIdevice.CloseConnection ();

    %% If you want to unload the dll and destroy the class object
    Controller.Destroy;
    clear Controller;
    clear PIdevice;
catch
    Controller.Destroy;
    clear Controller;
    clear PIdevice;
    rethrow(lasterror);
end